server/src/main/java/org/opensearch/snapshots/SnapshotsService.java

```diff
@@ -247,10 +266,36 @@ public SnapshotsService(
      * @param listener snapshot completion listener
      */
     public void executeSnapshot(final CreateSnapshotRequest request, final ActionListener<SnapshotInfo> listener) {
-        createSnapshot(
-            request,
-            ActionListener.wrap(snapshot -> addListener(snapshot, ActionListener.map(listener, Tuple::v2)), listener::onFailure)
-        );
+        startCreateSnapshot(request, listener);
+    }
+
+    /**
+     * This method calls {@link #createSnapshot(CreateSnapshotRequest, ActionListener)} to create snapshot if snapshot
+     * V2 is not enabled.
+     * For V2 enabled snapshots, {@link #createSnapshotV2(CreateSnapshotRequest, ActionListener)} is called and
+     * appropriate listeners are mapped
+     * @param request snapshot request
+     * @param listener snapshot completion listener
+     */
+    public void startCreateSnapshot(final CreateSnapshotRequest request, final ActionListener<SnapshotInfo> listener) {
```

Comment:
Please verify if `request.waitForCompletion()` is set to true. In that case we add a listener to `snapshotCompletionListeners` and it is signaled post the the end of snapshot . The right way might be wrap this listener if snapshot  v2 is enabled .

Reply:
The completion listener is added in `addListener` call`https://github.com/opensearch-project/OpenSearch/blob/6504169602c532a931a3c80e31857558c9b2a4fc/server/src/main/java/org/opensearch/snapshots/SnapshotsService.java#L300`. Since snapshotV2 is sync, I tried to avoid using this completion listener.